source/agora/node/Validator.d

```diff
@@ -411,4 +416,50 @@ public class Validator : FullNode, API
         super.shutdown();
         this.nominator.storeLatestState();
     }
+
+    /***************************************************************************
+
+        Check if the current enrollment is about to expire or validator is not
+        enrolled. Enroll when necessary.
+
+        Params:
+            block_height = Current block height
+
+    ***************************************************************************/
+
+    private void checkAndEnroll (Height block_height) @safe
+    {
+        Hash enroll_key = this.enroll_man.getEnrolledUTXO(this.utxo_set.getUTXOFinder());
+
+        if (enroll_key == Hash.init && (enroll_key = this.getFrozenUTXO()) == Hash.init)
+            return; // Not enrolled and no frozen UTXO
+
+        const enrolled = this.enroll_man.getEnrolledHeight(enroll_key);
+
+        // This validators enrollment will expire next cycle or not enrolled at all
+        if (enrolled == ulong.max || block_height + 1 >= enrolled + this.params.ValidatorCycle)
+            this.network.sendEnrollment(this.enroll_man.createEnrollment(enroll_key));
```

Comment:
I think it would be useful to log this before sending the enrollment.

Reply:
Would that be a `log.trace` or a higher level?

Reply:
trace is fine!